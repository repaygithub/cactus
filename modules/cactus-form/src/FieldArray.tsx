import { useForm } from 'react-final-form'
import generateKey from 'helpers/generateKey'
import React from 'react'

const getKey = obj => (obj?.key !== undefined ? obj.key : obj)

const compareKeys = (left, right) => {
  const length = left?.length || 0
  if (length !== right?.length) return false
  for (let i = 0; i < length; i++) {
    if (getKey(left[i]) !== getKey(right[i])) return false
  }
  return true
}

const FieldArray = ({ name, component, subscription, ...rest }) => {
  const form = useForm()
  const ref = React.useRef()
  const mutators = React.useMemo(
    () =>
      Object.keys(form.mutators).reduce((result, key) => {
        result[key] = (...args) => form.mutators[key](name, ...args)
        return result
      }, {}),
    [form.mutators, name]
  )
  const trigger = React.useReducer(x => !x, false)[1]
  React.useEffect(() => {
    let keys = []
    return form.registerField(
      name,
      state => {
        ref.current = state.value
        if (state.value && getKey && setKey) {
          const newKeys = []
          const changed = state.length !== keys.length || state.value.reduce(
            (hasChanged, arrayVal, index) => {
              const key = getKey(arrayVal) ?? setKey(arrayVal)
              newKeys.push(key)
              return hasChanged || key !== keys[index]
            },
            false,
          )
          keys = newKeys
          if (changed) trigger()
        }
      },
      { ...subscription, value: true, length: true }
    )
  }, [form, name, subscription, trigger])

  // I wouldn't use this interface normally, but I want it to be compatible for now.
  const length = ref.current?.length || 0
  const props = {
    ...rest,
    meta: {},
    fields: {
      name,
      length,
      value: ref.current,
      map: fn => {
        const results = []
        for (let i = 0; i < length; i++) {
          results.push(fn(`${name}[${i}]`, i))
        }
        return results
      },
      ...mutators,
    },
  }
  return React.createElement(component, props)
}

export default FieldArray
