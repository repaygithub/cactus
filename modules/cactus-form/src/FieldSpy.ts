import { FieldState, FieldSubscription } from 'final-form'
import PropTypes from 'prop-types'
import React from 'react'
import { useForm } from 'react-final-form'

import { FC, RenderFunc, RenderProps, UnknownProps } from './types'

interface FieldSpyProps extends UnknownProps, RenderProps {
  fieldName: string
  subscription: FieldSubscription
  isEqual?: (left: any, right: any) => boolean
}

const FieldSpy: FC<FieldSpyProps> = ({
  fieldName,
  subscription,
  component,
  render,
  isEqual,
  ...props
}) => {
  if (typeof props.children === 'function') {
    render = props.children as RenderFunc
    delete props.children
  }
  const form = useForm()
  const [state, setState] = React.useState<FieldState<unknown>>()
  React.useEffect(() => {
    const config = isEqual && { isEqual }
    return form.registerField(fieldName, setState, subscription, config)
  }, [fieldName, form, subscription, isEqual])
  if (!state) return null
  Object.assign(props, state)
  if (render) {
    return render(props)
  } else if (component) {
    return React.createElement(component, props)
  }
  return null
}
FieldSpy.propTypes = {
  fieldName: PropTypes.string.isRequired,
  subscription: PropTypes.objectOf(PropTypes.bool).isRequired,
}
export default FieldSpy
