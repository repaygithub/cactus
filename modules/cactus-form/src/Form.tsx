import { formSubscriptionItems } from 'final-form'
import React from 'react'
import { Form as FinalForm, FormProps as FinalFormProps, FormRenderProps } from 'react-final-form'

function renderSimpleForm<FormValues>({
  handleSubmit,
  form,
  ...rest
}: FormRenderProps<FormValues>) {
  // `react-final-form` adds all the state properties even if they don't have values.
  for (const key of formSubscriptionItems) {
    // @ts-ignore
    if (rest[key] === undefined) delete rest[key]
  }
  return <form onSubmit={handleSubmit} onReset={() => form.restart()} {...rest} />
}

function Form<FormValues>(props: FinalFormProps<FormValues>) {
  const component = props.component || props.as || 'form'
  if (component === 'form' && !props.render) {
    props = { ...props, render: renderSimpleForm }
    delete props.component
  }
  console.log(props.subscription)
  return FinalForm(props)
}
Form.defaultProps = { subscription: {} }

export default Form
