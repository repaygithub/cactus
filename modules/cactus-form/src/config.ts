import { ConfigurableComponent, RenderFunc } from './types'

const makeConfigurableComponent = <P>(
  fn: RenderFunc<P>,
  defaults: Partial<P>
): ConfigurableComponent<P> => {
  const component = fn as ConfigurableComponent<P>
  component.defaultProps = defaults
  component.withDefaults = (newDefaults: Partial<P>) => {
    const clone: ConfigurableComponent<P> = component.bind(null) as any
    Object.assign(clone, component)
    return makeConfigurableComponent<P>(clone, newDefaults)
  }
  component.configureDefaults = (newDefaults: Partial<P>) => {
    const prevDefaults = { ...component.defaultProps }
    component.defaultProps = newDefaults
    return prevDefaults
  }
  component.initialDefaults = () => defaults
  return component
}
export default makeConfigurableComponent
