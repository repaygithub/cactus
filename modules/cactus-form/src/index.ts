export { default as DependentField } from './DependentField'
export { default as FieldSpy } from './FieldSpy'
export { default as Field, validateRequired } from './Field'
