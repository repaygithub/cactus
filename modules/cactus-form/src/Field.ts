import {
  CheckBoxField,
  DateInputField,
  FileInputField,
  RadioButtonField,
  SelectField,
  TextInputField,
  ToggleField,
} from '@repay/cactus-web'
import { FieldSubscription } from 'final-form'
import PropTypes from 'prop-types'
import React from 'react'
import { FieldMetaState, useField, UseFieldConfig } from 'react-final-form'

import { makeConfigurableComponent } from './config'
import { RenderFunc, RenderProps, UnknownProps } from './types'

interface FieldConfig extends UseFieldConfig<unknown> {
  component?: React.ElementType<any>
}

const FIELD_PROPS: (keyof FieldConfig)[] = [
  'afterSubmit',
  'allowNull',
  'beforeSubmit',
  'data',
  'defaultValue',
  'format',
  'formatOnBlur',
  'initialValue',
  'isEqual',
  'multiple',
  'parse',
  'type',
  'validate',
  'validateFields',
  'value',
]

type ComponentMapper = (props: UnknownProps) => React.ElementType<any>
type PostProcessor = (props: UnknownProps, meta: FieldMetaState<unknown>) => UnknownProps

export interface FieldProps extends UnknownProps, RenderProps, FieldConfig {
  name: string
  getFieldComponent?: ComponentMapper
  processMeta?: PostProcessor
}

const DEFAULT_SUB: FieldSubscription = {
  value: true,
  error: true,
  touched: true,
  submitError: true,
  modifiedSinceLastSubmit: true as any,
}

const addError: PostProcessor = (props, meta) => {
  const error = (meta.touched && meta.error) || (!meta.modifiedSinceLastSubmit && meta.submitError)
  if (error && !props.error) {
    props.error = error
  }
  return props
}

const defaultFieldComponent: ComponentMapper = (props) => {
  const fieldType = popAttr(props, 'fieldType') || props.type
  switch (fieldType) {
    case 'boolean':
      props.type = 'checkbox'
      return ToggleField

    case 'checkbox':
      props.type = 'checkbox'
      return CheckBoxField

    case 'radio':
      props.type = 'radio'
      return RadioButtonField

    case 'select_multiple':
    case 'select-multiple':
    case 'multiSelect':
      props.multiple = true
      return SelectField
    case 'creatableDropdown':
    case 'select':
      props.comboBox = fieldType === 'creatableDropdown'
      return SelectField

    case 'date':
    case 'time':
    case 'datetime':
      props.type = fieldType
      return DateInputField

    case 'file':
      return FileInputField

    case 'hidden':
      props.type = fieldType
      return 'input'

    case 'number':
    case 'numeric':
      props.type = 'number'
      return TextInputField

    default:
      if (props.options) {
        return SelectField
      }
      return TextInputField
  }
}

const popAttr = (obj: UnknownProps, attr: string) => {
  const val = obj[attr]
  delete obj[attr]
  return val
}

const Field: RenderFunc<FieldProps> = ({
  name,
  render,
  component,
  subscription = DEFAULT_SUB,
  processMeta = addError,
  getFieldComponent = defaultFieldComponent,
  ...props
}) => {
  if (typeof props.children === 'function') {
    render = popAttr(props, 'children') as RenderFunc
  } else if (!component && !render) {
    if (props['as']) {
      component = popAttr(props, 'as') as React.ElementType<any>
    } else {
      component = getFieldComponent(props)
    }
  }
  const fieldConfig: FieldConfig = { component, subscription }
  // This removes `type` && `multiple` from props, but they're re-added by `useField`.
  for (const key of FIELD_PROPS) {
    if (key in props) {
      fieldConfig[key] = popAttr(props, key) as any
    } else if (component && (component as any)[key]) {
      // @ts-ignore For setting component defaults, useful for `format`/`parse`/`validate`.
      fieldConfig[key] = component[key]
    }
  }
  if (props.required && !fieldConfig.validate) {
    fieldConfig.validate = validateRequired
  }

  const field = useField(name, fieldConfig)
  const inputProps = Object.assign(field.input, props)
  const fieldProps = processMeta(inputProps, field.meta)
  if (render) {
    return render(fieldProps)
  } else if (component) {
    return React.createElement(component, fieldProps)
  }
  return null
}

;(Field as any).propTypes = { name: PropTypes.string.isRequired }
export default makeConfigurableComponent(Field, {
  subscription: DEFAULT_SUB,
  getFieldComponent: defaultFieldComponent,
  processMeta: addError,
})

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const validateRequired = (value: any): string | undefined => {
  if ((!value && value !== 0 && value !== false) || value?.length === 0) {
    return 'Missing required field'
  }
}
