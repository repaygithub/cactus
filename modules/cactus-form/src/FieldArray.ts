import { ARRAY_ERROR, fieldSubscriptionItems, FieldConfig } from 'final-form'
import { useForm } from 'react-final-form'
import React from 'react'

import { popAttr, getFieldConfig } from './helpers'
import { makeConfigurableComponent } from './config'

type FieldArrayConfig = FieldConfig<unknown[]>

const generateKey = (prefix: string) => Math.random().toString().replace(/^0/, prefix)

const simpleKeyFunc = (obj: any) => {
  if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
    return obj.hasOwnProperty('key') ? obj.key : (obj.key = generateKey('array'))
  }
}

const DEFAULTS = { value: [], length: 0 }

const ARRAY_PROPS: (keyof FieldArrayConfig)[] = [
  'afterSubmit',
  'beforeSubmit',
  'data',
  'defaultValue',
  'initialValue',
  'isEqual',
  'validateFields',
]

const processState = (fieldState, prevState, subKeys, keyFunc) => {
  const nextState = {}
  let changed = false
  if (keyFunc) {
    const prevKeys = prevState.keys as React.Key[]
    const nextKeys = []
    changed = (fieldState.value || []).reduce(
      (hasChanged, val, index) => {
        const key = keyFunc(val)
        nextKeys.push(key)
        return hasChanged || key !== prevKeys[index]
      },
      !prevKeys,
    )
    nextState.keys = changed ? nextKeys : prevKeys
  }
  for (const key of subKeys) {
    let val = fieldState[key] ?? DEFAULTS[key]
    if ((key === 'error' || key === 'submitError') && Array.isArray(val)) {
      val = val[ARRAY_ERROR]
    }
    nextState[key] = val
    changed = changed || prevState[key] !== val
  }
  return changed ? nextState : prevState
}

const processArrayError = (error: any, checkForPromise: boolean = false) => {
  if (!error || Array.isArray(error)) {
    return error
  } else if (checkForPromise && typeof error.then === 'function') {
    return error.then(processArrayError)
  } else {
    const placeholder: any = []
    placeholder[ARRAY_ERROR] = error
    return placeholder
  }
}

// TODO No way to fully automate, would always need to be called in `onSubmit`
const removeKeys = (values, formApi) => {
  const cleanupFuncs = weakMap.get(formApi) || {}
  return runCleanup(values, cleanupFuncs)
}

const simpleCleanup = (value) => {
  if (value && typeof value === 'object' && value.hasOwnProperty('key')) {
    const { key, ...copy } = value
    return copy
  }
  return value
}

const runCleanup = (values, cleanupFuncs) => {
  if (typeof cleanupFuncs === 'function') {
    return cleanupFuncs(values)
  } else if (!values || !cleanupFuncs) {
    return values
  } else if (Array.isArray(cleanupFuncs)) {
    return values.map((val, i) => runCleanup(val, cleanupFuncs[i]))
  } else {
    const copy = {}
    for (const key of Object.keys(values)) {
      copy[key] = runCleanup(values[key], cleanupFuncs[key])
    }
    return copy
  }
}

const FieldArray = ({ name, component, render, subscription, keyFunc, processMeta, validate, ...rest }) => {
  if (typeof rest.children === 'function') {
    render = popAttr(rest, 'children') as RenderFunc
  } else if (!component && !render && rest['as']) {
    component = popAttr(rest, 'as') as React.ElementType<any>
  }
  const form = useForm()

  const ref = React.useRef()
  if (ref.current?.name !== name) {
    const mutators = { change: (value?: unknown[]) => form.change(name, value) }
    for (const key of Object.keys(form.mutators)) {
      const mutator = form.mutators[key]
      mutators[key] = (...args) => mutator(name, ...args)
    }
    ref.current = { name, mutators }
  }
  ref.current.validate = validate

  const fieldConfig = getFieldConfig<FieldArrayConfig>(ARRAY_PROPS, rest, component)
  const register = (callback) => {
    // There's basically no reason to ever not subscribe to length.
    const sub = { ...subscription, length: true }
    const subKeys = fieldSubscriptionItems.filter((k) => sub[k])
    sub.value = sub.value || !!keyFunc
    const onStateChange = callback
      ? (state) => callback(processState(state, {}, subKeys, keyFunc))
      : (state) => setState((old) => processState(state, old, subKeys, keyFunc))
    fieldConfig.silent = !!callback
    const validator = (value, formValues, meta) => {
      const error = ref.current.validate?.(value, formValues, meta)
      return processArrayError(error, true)
    }
    fieldConfig.getValidator = () => validator
    return form.registerField(name, onStateChange, sub, fieldConfig)
  }

  const [fieldState, setState] = React.useState(() => {
    let state = undefined
    register((s) => (state = s))()
    return state
  })
  React.useEffect(register, [name, form])

  const length = fieldState.length
  const arrayProps = {
    ...ref.current.mutators,
    map: (fn, thisArg) => {
      if (thisArg !== undefined) fn = fn.bind(thisArg)
      const results = []
      for (let i = 0; i < length; i++) {
        results.push(fn(`${name}[${i}]`, i))
      }
      return results
    },
    ...fieldState,
    name,
  }
  const props = processMeta(rest, arrayProps)
  if (render) {
    return render(props)
  } else if (component) {
    return React.createElement(component, props)
  }
  return null
}

export default makeConfigurableComponent(FieldArray, { keyFunc: simpleKeyFunc, processMeta: (props, meta) => Object.assign(meta, props) })
