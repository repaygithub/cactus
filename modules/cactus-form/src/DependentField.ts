import { FieldState, FieldSubscription } from 'final-form'
import React from 'react'
import { useForm } from 'react-final-form'

import { makeConfigurableComponent } from './config'
import Field, { FieldProps } from './Field'
import { RenderFunc, UnknownProps } from './types'

type DependencyChangeHandler = (state: FieldState<unknown>, props: UnknownProps) => void

interface DependencyConfig extends FieldSubscription {
  onChange?: DependencyChangeHandler
}

interface Box {
  onChange: DependencyChangeHandler
  props: UnknownProps
}

interface DependentFieldProps extends FieldProps {
  dependsOn: string | string[] | Record<string, DependencyConfig>
  onDependencyChange?: DependencyChangeHandler
}

const DEFAULT_SUB: FieldSubscription = { value: true }
const noop = () => undefined

const DependentField: RenderFunc<DependentFieldProps> = ({
  dependsOn,
  onDependencyChange = noop,
  ...props
}) => {
  const field = Field(props)
  const box = React.useRef({} as Box).current
  box.onChange = onDependencyChange
  box.props = field?.props || {}

  const form = useForm('DependentField')
  React.useEffect(() => {
    let initialized = false
    let unsubs: (() => void)[]
    const makeHandler =
      (config: DependencyConfig = {}) =>
      (state: FieldState<unknown>) => {
        if (initialized) {
          const { onChange = box.onChange } = config
          onChange(state, box.props)
        }
      }
    if (typeof dependsOn === 'string') {
      unsubs = [form.registerField(dependsOn, makeHandler(), DEFAULT_SUB)]
    } else if (Array.isArray(dependsOn)) {
      const update = makeHandler()
      unsubs = dependsOn.map((fieldName) => form.registerField(fieldName, update, DEFAULT_SUB))
    } else {
      unsubs = Object.keys(dependsOn).map((fieldName) => {
        const config = dependsOn[fieldName]
        return form.registerField(fieldName, makeHandler(config), config)
      })
    }
    initialized = true
    return () => unsubs.forEach((unsub) => unsub())
  }, [dependsOn, box, form])
  return field
}
export default makeConfigurableComponent(DependentField, {})
