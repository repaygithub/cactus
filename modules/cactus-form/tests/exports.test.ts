import * as cactusForm from '../src/index'

describe('@repay/cactus-form', (): void => {
  test('should match snapshot of exports', (): void => {
    expect(Object.keys(cactusForm)).toMatchSnapshot()
  })
})
