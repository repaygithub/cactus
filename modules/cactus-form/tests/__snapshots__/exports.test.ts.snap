// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@repay/cactus-form should match snapshot of exports 1`] = `
Array [
  "ARRAY_ERROR",
  "DependentField",
  "FORM_ERROR",
  "Field",
  "FieldSpy",
  "Form",
  "FormSpy",
  "SubmitButton",
  "configOptions",
  "createForm",
  "fieldSubscriptionItems",
  "formSubscriptionItems",
  "useField",
  "useForm",
  "useFormState",
  "validateRequired",
  "withTypes",
]
`;
